### PR TITLE
gh-110913: Fix WindowsConsoleIO chunking of UTF-8 text

### PR DESCRIPTION
**Pull request by sorgloomer**
Fix the loop that searches for an UTF-8 sequence boundary

 * Issue: https://github.com/python/cpython/issues/110913

* Issue: gh-110913

### CODE REVIEW

**Comment by zooba** [comment]
I think we should have a NEWS entry for this, just to acknowledge that we changed the fix. Something like `WindowsConsoleIO now correctly chunks large buffers without splitting up UTF-8 sequences`

The change looks fine to me, but I'd like one more ACK before merging in case my head still isn't clear.

**Comment by sorgloomer** [comment]
Thanks! Added the NEWS entry.

File: Modules/_io/winconsoleio.c [hunk 1]
```diff
@@ -134,6 +134,18 @@ char _PyIO_get_console_type(PyObject *path_or_fd) {
...
+    DWORD count = 1;
+    if (len == 0 || (buf[len - 1] & 0x80) == 0)
+        return len;
+    for (;; count++) {
+        if (count > 3 || count >= len)
+            return len;
+        if ((buf[len - count] & 0xc0) != 0x80)
+            return len - count;
+    }
+}
```
Comment (serhiy-storchaka) [inline Modules/_io/winconsoleio.c]:
```suggestion
static DWORD
_find_last_utf8_boundary(const char *buf, DWORD len)
{
    /* This function never returns 0, returns the original len instead */
    DWORD count = 1;
    if (len == 0 || (buf[len - 1] & 0x80) == 0) {
        return len;
    }
    for (;; count++) {
        if (count > 3 || count >= len) {
            return len;
        }
        if ((buf[len - count] & 0xc0) != 0x80) {
            return len - count;
        }
    }
}
```

File: Misc/NEWS.d/next/Windows/2023-10-19-21-46-18.gh-issue-110913.CWlPfg.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+WindowsConsoleIO now correctly chunks large buffers without splitting up UTF-8 sequences
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Windows/2023-10-19-21-46-18.gh-issue-110913.CWlPfg.rst]:
```suggestion
WindowsConsoleIO now correctly chunks large buffers without splitting up UTF-8 sequences.
```

**Review by serhiy-storchaka — COMMENTED** [review]
Is it even possible to write tests for this?

**Comment by zooba** [comment]
> Is it even possible to write tests for this?

Not without refactoring the entire thing (e.g. to make the actual `WriteConsoleW` call encoding-unaware and create a specific TextIOWrapper just for this case that does the chunking). Or to make it possible to require a WindowsConsoleIO instance to talk to a file (which then could not use `WriteConsoleW` I believe, though it may be possible to trick it).

It's a harmless enough issue that I think we're okay. Misunderstanding the UTF-8 format is what the problem was this time, and that's been triple checked now.

**Comment by serhiy-storchaka** [comment]
I'm just wondering if it's possible to get invalid UTF-8 here (using "surrogateescape"?). If yes, then we should use more sophisticated way to find the longest valid prefix of invalid sequence to meet recomendations for handling invalid UTF-8. Otherwise LGTM.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by zooba** [comment]
If it's undecodable by `MultiByteToWideChar` at this point, we'll have raised an error and never entered into this branch (`wlen` will be 0). And even if it did, it wouldn't be rendered correctly anyway and we aren't really making things worse by rendering it slightly more incorrectly.

Thanks for the review!

**Pull request merged**